Add an experimental tree DDS (#4821)
microsoft/FluidFramework: experimental/dds/tree/README.md

```typescript
tree.editor.insert(fooNode, StablePlace.atStartOf({ parent: initialTree.identifier, label: 'foo' }));
tree.editor.insert(barNode, StablePlace.atStartOf({ parent: initialTree.identifier, label: 'bar' }));
```

This would insert `fooNode` at the start of the "foo" trait underneath the initial tree's root node, and `barNode` underneath the "bar" trait.
Each operation would be performed in its own `Edit`, which is `SharedTree`'s transactional atom.

If it is undesirable that one of the above operations could fail to apply while the other could succeed, you should instead leverage `Checkout`. A `Checkout`--whose name is inspired from source control--can be thought of as a local view of the `SharedTree` which provides [snapshot isolation](https://en.wikipedia.org/wiki/Snapshot_isolation#:~:text=In%20databases%2C%20and%20transaction%20processing,the%20transaction%20itself%20will%20successfully). Editing the `SharedTree` using a `Checkout` can be done by opening an edit, applying a number of changes, and closing the edit.

```typescript

> const checkout = new BasicCheckout(tree);

Author: I considered including a reference to the [atomicity wikipedia page](https://en.wikipedia.org/wiki/Atomicity_(database_systems)) on the reference to "transactional atom" above. Do you think that would be useful to answer your question? Or are you referring to some other concern?

checkout.openEdit();
checkout.applyChanges(Insert.create([fooNode], StablePlace.atStartOf({ parent: initialTree, label: 'foo' })));
checkout.applyChanges(Insert.create([barNode], StablePlace.atStartOf({ parent: initialTree, label: 'bar' })));
checkout.closeEdit();
```

"Move" and "delete" operations have the added complexity of needing to specify locations (`StableRange`s) within the `SharedTree` which should be moved (or deleted, respectively). A `StableRange` consists of a start `StablePlace` and an end `StablePlace`.
`StablePlace`s are not nodes, but instead places where nodes could be inserted. Each place consists of an "anchor," which is either a trait or another node.

> Say we wanted to delete the `fooNode` we inserted above. There are 4 ways we could specify the `StableRange` to delete which are all equivalent in the absence of collaboration:

Author: Yes, I like that better.

```suggestion
Say we wanted to delete the `fooNode` we inserted above. There are 4 ways we could specify the `StableRange` to delete which are all equivalent in the absence of concurrent editing:
```

```typescript
const trait = { parent: initialTree, label: 'foo' };
const stableRange1 = StableRange.from(StablePlace.atStartOf(trait)).to(StablePlace.atEndOf(trait));
const stableRange2 = StableRange.from(StablePlace.atStartOf(trait)).to(StablePlace.after(fooNode));
const stableRange3 = StableRange.from(StablePlace.before(fooNode)).to(StablePlace.atEndOf(trait));
const stableRange4 = StableRange.from(StablePlace.before(fooNode)).to(StablePlace.after(fooNode));
```

> Once collaboration is considered, the different ways to anchor this `StableRange` may impact whether or not this edit conflicts with others.

Author: ```suggestion
Once concurrent edits are considered, the different ways to anchor this `StableRange` may impact whether or not this edit conflicts with others.
```

Also note that there are some more convenient shorthands for several of these specifications. See `StableRange` documentation for more information.

# Status

SharedTree is in active, but still relatively early development. As such, it is lacking in some areas (such as performance testing). For an idea of some future features we'd like to support, see [Future Plans](./docs/Future.md).

Implementation-wise:

-   Document format may change only in major releases, and SharedTree is commited to backwards compatibility (support for older documents). For more information on this commitment, see the notes in [PersistedTypes.ts](./src/PersistedTypes.ts).
-   APIs are not yet stable, and those beyond what's needed for the MVP (ex: history editing and inspection) are not provided yet. Core APIs are not likely to significantly change.
-   Performance is generally reasonable. However, this assessment was made using integration-style performance tests of consuming applications. Though it's on the roadmap, there are currently no rigorous performance tests which are isolated to SharedTree.

Design wise:

> -   SharedTree is always created with an uninitialized state. It is up to the application to use the shared tree to initialize the tree to something else if needed.

Author: Yeah. I believe the intent of the sentence is to convey the idea that there isn't some analog of the `Map` constructor that takes in a list of key-value pairs.

-   There are still open questions regarding how SharedTree will relate to the rest of the Fluid ecosystem.
    For example, we do not have suggested design patterns for when users of SharedTree should used references to other Fluid components versus including the data for children as subtrees.

# Edits

An `Edit` is the basic unit of transactionality in `SharedTree`. It specifies how to modify a document via a sequence of changes (see [PersistedTypes.ts](.\src\PersistedTypes.ts)). Each edit applied to a version of the document (a Snapshot) produces a new version of the document.

> Once an edit is acknowledged by Fluid (and thus it has a sequence number, and will be included in summaries), the version of the document it applies to is fixed: it will not be applied to any revision other than the one produced by its preceding edit, transitively ack to the initial tree.

Author: ```suggestion
Once an edit is acknowledged by the Fluid service (and thus it has a sequence number, and will be included in summaries), the version of the document it applies to is fixed: it will not be applied to any revision other than the one produced by its preceding edit.
```

(resolving other comment as well with this suggestion)

> Once an edit is acknowledged by Fluid (and thus it has a sequence number, and will be included in summaries), the version of the document it applies to is fixed: it will not be applied to any revision other than the one produced by its preceding edit, transitively ack to the initial tree.

Author: Yeah, I believe "transitively ..." should just be omitted.

There may be operations that will create new edits based on existing ones and apply them in a different context, but these are logically considered new edits.

## Conflicts

Due to the collaborative and distributed nature of SharedTree,

> Changes may be constructed based on a version of the tree that differs from the one they end up getting applied to.

Author: Either makes sense, but I think you're right that `Edit` is probably better--there is a more direct API analog to it.
```suggestion
Edits may be constructed based on a version of the tree that differs from the one they end up getting applied to.
```
